### ISSUE TITLE
Sync API for HttpClient

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#32125]
Provide sync API on `HttpClient`, which currently has only async methods.

## Motivation

Azure SDK currently exposes synchronous API which does sync-over-async on `HttpClient` (see [HttpClientTransport.cs](https://github.com/Azure/azure-sdk-for-net/blob/8902f9fe785ba162ee3b5fca9dee4bc60d937081/sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs#L53)). It would be much better if we provided them with sync API with proper sync implementation, where feasible.

Also there are lots of existing code bases that have very deep synchronous call stacks. Developers are simply not willing to rewrite these code bases to be asynchronous. If they need to call async only API in these synchronous methods, they use sync-over-async, which then in turn causes "soft" deadlock. We want to provide synchronous APIs for these developers because synchronous APIs, although inefficient, can can help in avoiding these soft deadlocks.

Another advantage of sync API is that it is much easier to grasp. Especially for people with no prior knowledge of asynchronous processing. If someone is starting with `HttpClient`, they also have to be knowledgeable of C# `async/await`. Also many examples with ``HttpClient` might be simplified and thus making the entry into .NET easier.

### Minimal Necessary Change

This change is based on @stephentoub prototype in https://github.com/stephentoub/corefx/commit/0e4d6401e76c79f8ef81b5064322b8de18a115a2. The prototype introduces synchronous API for `SocketsHttpHandler` and also properly implements it for HTTP 1.1 scenarios (for HTTP 2 does sync-over-async). This proposal extends the existing prototype with sync API on `HttpClient`. Following changes are a minimal set to achieve synchronous `HttpClient.Send` behaving trully synchronously at least for HTTP 1.1. 

Note that `CancellationToken` is used in synchronous methods in order to propagate `HttpClient` timeout and cancellations. For `HttpContent` it's up for discussion whether to add other overload to `CopyTo` and `SerializeToStream` to match the async counterparts. 

The prototype/early implementation for the minimal change can be found in my branch https://github.com/ManickaP/runtime/tree/mapichov/sync_http_api.
``` c#
// The main classes where we want the sync API.
public partial class HttpMessageInvoker : IDisposable
{
    ...
    // Existing async method
    public virtual Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);

    // Proposed new sync methods
    public virtual HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}
public partial class HttpClient : HttpMessageInvoker
{
    ...
    // Existing async methods
    public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request);
    public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption);
    public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken);
    public override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
    
    // Proposed new sync methods
    public HttpResponseMessage Send(HttpRequestMessage request, HttpCompletionOption completionOption = default, CancellationToken cancellationToken = default);
    public override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}

// The changes bellow enable proper implementation of the sync API.

// HttpMessageHandler and derived classes
public abstract partial class HttpMessageHandler : IDisposable
{
    ...
    // Existing async method
    protected internal abstract Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
    
    // Proposed new sync method
    protected internal virtual HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}
public abstract partial class DelegatingHandler : HttpMessageHandler
{
    ...
    // Proposed new sync method
    protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}
public abstract partial class HttpClientHandler : HttpMessageHandler
{
    ...
    // Proposed new sync method
    protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}
public abstract partial class SocketsHttpHandler : HttpMessageHandler
{
    ...
    // Proposed new sync method
    protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}
public abstract partial class MessageProcessingHandler : DelegatingHandler
{
    ...
    // Proposed new sync method
    protected internal sealed override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
}

// HttpContent and derived classes
public abstract partial class HttpContent : IDisposable
{
    ...
    // Existing async methods
    public Task CopyToAsync(Stream stream);
    public Task CopyToAsync(Stream stream, TransportContext context);
    public Task CopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);
    public Task CopyToAsync(Stream stream, CancellationToken cancellationToken);

    protected abstract Task SerializeToStreamAsync(Stream stream, TransportContext context);
    protected virtual Task SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);

    // Proposed new sync methods
    public void CopyTo(Stream stream, TransportContext context, CancellationToken cancellationToken);

    protected virtual void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
public partial class ByteArrayContent : HttpContent
{
    ...
    // Proposed new sync method
    protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
public partial class MultipartContent : HttpContent, IEnumerable<HttpContent>, IEnumerable
{
    ...
    // Proposed new sync method
    protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
public sealed partial class ReadOnlyMemoryContent : HttpContent
{
    ...
    // Proposed new sync method
    protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
public partial class StreamContent : HttpContent
{
    ...
    // Proposed new sync method
    protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
```

### Full Change

This includes sync counterparts to all async `HttpClient` methods. Since their implementation delegates to `Send` underneath, there's no need to add any other supporting sync methods.
Note that all the sync methods on `HttpClient` (except for `HttpMessageInvoker.Send` override) do not need oveloads with `CancellationToken`. Whether to include them or not is up for discussion.
``` c#
public partial class HttpClient : HttpMessageInvoker
{
    ...
    // Existing async methods
    public Task<HttpResponseMessage> DeleteAsync(string requestUri);
    public Task<HttpResponseMessage> DeleteAsync(string requestUri, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> DeleteAsync(Uri requestUri);
    public Task<HttpResponseMessage> DeleteAsync(Uri requestUri, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> GetAsync(string requestUri);
    public Task<HttpResponseMessage> GetAsync(string requestUri, HttpCompletionOption completionOption);
    public Task<HttpResponseMessage> GetAsync(string requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> GetAsync(string requestUri, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> GetAsync(Uri requestUri);
    public Task<HttpResponseMessage> GetAsync(Uri requestUri, HttpCompletionOption completionOption);
    public Task<HttpResponseMessage> GetAsync(Uri requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> GetAsync(Uri requestUri, CancellationToken cancellationToken);
    public Task<byte[]> GetByteArrayAsync(string requestUri);
    public Task<byte[]> GetByteArrayAsync(string requestUri, CancellationToken cancellationToken);
    public Task<byte[]> GetByteArrayAsync(Uri requestUri);
    public Task<byte[]> GetByteArrayAsync(Uri requestUri, CancellationToken cancellationToken);
    public Task<Stream> GetStreamAsync(string requestUri);
    public Task<Stream> GetStreamAsync(string requestUri, CancellationToken cancellationToken);
    public Task<Stream> GetStreamAsync(Uri requestUri);
    public Task<Stream> GetStreamAsync(Uri requestUri, CancellationToken cancellationToken);
    public Task<string> GetStringAsync(string requestUri);
    public Task<string> GetStringAsync(string requestUri, CancellationToken cancellationToken);
    public Task<string> GetStringAsync(Uri requestUri);
    public Task<string> GetStringAsync(Uri requestUri, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content);
    public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content);
    public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> PostAsync(string requestUri, HttpContent content);
    public Task<HttpResponseMessage> PostAsync(string requestUri, HttpContent content, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> PostAsync(Uri requestUri, HttpContent content);
    public Task<HttpResponseMessage> PostAsync(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> PutAsync(string requestUri, HttpContent content);
    public Task<HttpResponseMessage> PutAsync(string requestUri, HttpContent content, CancellationToken cancellationToken);
    public Task<HttpResponseMessage> PutAsync(Uri requestUri, HttpContent content);
    public Task<HttpResponseMessage> PutAsync(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
    
    // Proposed new sync methods
    public HttpResponseMessage Delete(string requestUri);
    public HttpResponseMessage Delete(string requestUri, CancellationToken cancellationToken);
    public HttpResponseMessage Delete(Uri requestUri);
    public HttpResponseMessage Delete(Uri requestUri, CancellationToken cancellationToken);
    public HttpResponseMessage Get(string requestUri);
    public HttpResponseMessage Get(string requestUri, HttpCompletionOption completionOption);
    public HttpResponseMessage Get(string requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken);
    public HttpResponseMessage Get(string requestUri, CancellationToken cancellationToken);
    public HttpResponseMessage Get(Uri requestUri);
    public HttpResponseMessage Get(Uri requestUri, HttpCompletionOption completionOption);
    public HttpResponseMessage Get(Uri requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken);
    public HttpResponseMessage Get(Uri requestUri, CancellationToken cancellationToken);
    public byte[] GetByteArray(string requestUri);
    public byte[] GetByteArray(string requestUri, CancellationToken cancellationToken);
    public byte[] GetByteArray(Uri requestUri);
    public byte[] GetByteArray(Uri requestUri, CancellationToken cancellationToken);
    public Stream GetStream(string requestUri);
    public Stream GetStream(string requestUri, CancellationToken cancellationToken);
    public Stream GetStream(Uri requestUri);
    public Stream GetStream(Uri requestUri, CancellationToken cancellationToken);
    public string GetString(string requestUri);
    public string GetString(string requestUri, CancellationToken cancellationToken);
    public string GetString(Uri requestUri);
    public string GetString(Uri requestUri, CancellationToken cancellationToken);
    public HttpResponseMessage Patch(string requestUri, HttpContent content);
    public HttpResponseMessage Patch(string requestUri, HttpContent content, CancellationToken cancellationToken);
    public HttpResponseMessage Patch(Uri requestUri, HttpContent content);
    public HttpResponseMessage Patch(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
    public HttpResponseMessage Post(string requestUri, HttpContent content);
    public HttpResponseMessage Post(string requestUri, HttpContent content, CancellationToken cancellationToken);
    public HttpResponseMessage Post(Uri requestUri, HttpContent content);
    public HttpResponseMessage Post(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
    public HttpResponseMessage Put(string requestUri, HttpContent content);
    public HttpResponseMessage Put(string requestUri, HttpContent content, CancellationToken cancellationToken);
    public HttpResponseMessage Put(Uri requestUri, HttpContent content);
    public HttpResponseMessage Put(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
}
```

@stephentoub @KrzysztofCwalina @dotnet/ncl

**Comment by antonfirsov** [comment]
Do we have any existing synchronous API methods in .NET accepting `CancellationToken`?

**Comment by alnikola** [comment]
It would be better to clearly separate the public ```cancellationToken``` parameter purpose from HttpClient's own timeout which is set via ```Timeout```property. HttpClient's timeout will get a special handling logic once the PR #2281 is completed whereas the ```cancellationToken``` method parameter enables the caller to cancel request as they deem needed.

**Comment by davidsh** [comment]
>Do we have any existing synchronous API methods in .NET accepting CancellationToken?

We do not. CancellationToken's are something that was introduced into .NET as part of adding async/await/Task APIs.

In general, we have avoided adding new sync APIs for I/O. But as described above, there is reason now to support sync API for broadly used APIs like HttpClient.

I am curious, though, how the CancellationToken parameter will be used in these APIs. Sync APIs by their nature are difficult to cancel.

**Comment by Tratcher** [comment]
> I am curious, though, how the CancellationToken parameter will be used in these APIs. Sync APIs by their nature are difficult to cancel.

How is it any more difficult than implementing timeout support?

**Comment by scalablecory** [comment]
We will need to make the sync HTTP/1 implementation dispose the Socket, as the cancelable async methods won't be used. I think that's probably fine.

**Comment by Tratcher** [comment]
> We will need to make the sync HTTP/1 implementation dispose the Socket, as the cancelable async methods won't be used. I think that's probably fine.

That approach has a long history of AVs. Have those been dealt with?

**Comment by scalablecory** [comment]
> 
> 
> > We will need to make the sync HTTP/1 implementation dispose the Socket, as the cancelable async methods won't be used. I think that's probably fine.
> 
> That approach has a long history of AVs. Have those been dealt with?

Does it? I hope so!

**Comment by scalablecory** [comment]
We should consider using default parameters here:

```c#
public Task<HttpResponseMessage> GetAsync(string requestUri);
public Task<HttpResponseMessage> GetAsync(string requestUri, HttpCompletionOption completionOption);
public Task<HttpResponseMessage> GetAsync(string requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken);
public Task<HttpResponseMessage> GetAsync(string requestUri, CancellationToken cancellationToken);
```

into

```c#
public Task<HttpResponseMessage> GetAsync(string requestUri, HttpCompletionOption completionOption = HttpCompletionOption.ResponseContentRead, CancellationToken cancellationToken = default);
```

**Comment by ManickaP** [comment]
Add `CancelationToken`: we do not need to add all the `HttpClient` sync methods with it, we need only the `HttpMessageInvoker.Send` overload.

Also the `CancelationToken` handles not just timeout but request cancellation as well (`HttpClient.CancelPendingRequests`).

Finally, in the prototype, the `CancelationToken` is regularly checked for cancellation wherever it cannot be passed further down in the call chain. 

@alnikola: I've seen your PR, the timeout is still propagated by `CancelationToken` which would nicely work with the proposed solution.

**Comment by karelz** [comment]
Team discussion:
1. Use default values in `HttpClient.Send`
2. We propose to NOT take the convenience overloads ("full change") - @stephentoub did you have some arguments for taking them?
3. We are OK to take CancellationTokens for synchronous APIs - we should honor ALSO existing timeouts (e.g. `HttpClient.Timeout` and stream Timeouts)

**Comment by tmds** [comment]
Is the motivation strong enough for adding sync methods?

C# has great support for async programming.
And sync APIs means blocking threads...

**Comment by MihaMarkic** [comment]
One reason would be legacy code.

**Comment by Drawaes** [comment]
I think adding sync methods is a mistake and if upstream apis are exposing things as sync when there is underlying async that is their mistake to make, if you put these methods in I can see people falling into the trap of using them in a ecosystem that has embraced async. What is the scenario (outside of legacy) that you need a sync API these days. Servers are mostly async (Asp.net core being the most used in this space). UIs should definitely be using async

**Comment by benaadams** [comment]
They aren't sync; they are blocking apis; could they be prefixed or postfixed as such?
```csharp
 public HttpResponseMessage BlockingGet(string requestUri);
```
or
```csharp
 public HttpResponseMessage GetBlocking(string requestUri);
```

**Comment by stephentoub** [comment]
> What is the scenario (outside of legacy) that you need a sync API these days.

Two:
1. Simple use cases where the scalability of async isn't required, but more importantly...
2. Consuming APIs that are sync, and where forcing sync-over-async is worse than just sync. This isn't just "legacy", e.g. brand new Azure SDK clients exposing sync methods.

I'll let @Petermarcu and @KrzysztofCwalina speak to Azure SDK, though it's just one example.

**Comment by stephentoub** [comment]
> They aren't sync

How so?

**Comment by stephentoub** [comment]
> We propose to NOT take the convenience overloads ("full change") - @stephentoub did you have some arguments for taking them?

I'm fine with that.

**Comment by davidfowl** [comment]
How is HTTP/2 going to work?

**Comment by stephentoub** [comment]
> How is HTTP/2 going to work?

There will be some blocking involved; that's unavoidable. HTTP/1.x should be sync all the way down.

**Comment by Drawaes** [comment]
I guess my main issue is now the first thing people will see is the "Send" method and until now async only methods on HttpClient has been a pit of success story, I fear that will change.

**Comment by stephentoub** [comment]
> has been a pit of success story

That's exactly the problem... it hasn't been. Lots of call sites that require sync are forced to figure out how to block and then when they do they fall over at even small loads.

**Comment by stephentoub** [comment]
> the first thing people will see is the "Send" method

I don't think it is. Forget the alphabetical ordering, we're talking about 1 method on HttpClient among more than 30. GetAync/PostAsync are much more likely to be the "first thing people will see".

We're simply making it possible for sync clients to not suck.

**Comment by jkotas** [comment]
> We're simply making it possible for sync clients to not suck as much as they currently do.

This API is a performance optimization for a specific use. Do we have any estimates to quantify the improvement we expect to get out this API for these specific use cases, e.g. microbenchmark that demonstrates the improvement?

**Comment by stephentoub** [comment]
> This API is a performance optimization for a specific use. Do we have any estimates to quantify the improvement we expect to get out this API for these specific use cases, e.g. microbenchmark that demonstrates the improvement?

Yes. The key is it ends up requiring many more threads in order to sustain throughput. And it can take a long time for the thread pool to ramp up to the necessary level, while in the interim the system can essentially be deadlocked. That means either the dev needs to explicitly set a high min thread count to force the ramp up, or we need to make the pool way more aggressive at thread injection, which harms other cases.

**Comment by KrzysztofCwalina** [comment]
> I think adding sync methods is a mistake and if upstream apis are exposing things as sync when there is underlying async that is their mistake to make, if you put these methods in I can see people falling into the trap of using them in a ecosystem that has embraced async. 

@Drawaes, I totally agree with this problem. We want to solve it by having associated analyzers that detect when the sync APIs are called in a body of an async method

**Comment by jkotas** [comment]
> The key is it ends up requiring many more threads in order to sustain throughput.

Right, a net cost of waiting thread costs is ~10kB's of memory, let's make it 30kB. So this APIs allows you to save approximately 30kB * N memory where N is number of parallel pending network request. If this is important to save this, it sounds good to me. Just want to throw some numbers into the discussions - all discussions about performance have to have numbers :-)

> we need to make the pool way more aggressive at thread injection, which harms other cases.

We do not have to do this unconditionally. We can introduce an low-level API that let's threadpool know that the thread is stuck in sync-over-async and discount it in heuristics that decide whether to inject new threads. The advantage of doing this would be that it makes most sync-over-async problems better; the disadvantage is that you still burn 30kB of memory per instance so that it is not as good as the hand-crafted solution for the specific HttpClient case.

**Comment by stephentoub** [comment]
> a net cost of waiting thread costs

It's not that simple.  It's not just about the memory implications of threads sitting around waiting for something to do.  It's also about what happens when they do have something to do.  If I create a thousand threads just in case, and then a thousand work items get queued that end up doing computationally and memory-intensive work, I've now completely oversubscribed my system, just in case I needed to alleviate a potential deadlock.

This also requires the app developer to understand the implications of the libraries they're using, or the ones they're indirectly using, and configure the thread pool because some sync APIs are being used.

> We can introduce a low-level API that let's threadpool know that the thread is stuck in sync-over-async and discount it in heuristics that decide whether to inject new threads.

I'm happy with doing that as well, assuming we can come up with the right, well-tuned heurstics (I don't think it's as easy as just pretending that the thread currently used for sync-over-async doesn't exist).

**Comment by davidfowl** [comment]
> There will be some blocking involved; that's unavoidable. HTTP/1.x should be sync all the way down.

Same with HTTP/3. I don't have a big issue with this change I just think the people thinking about it are thinking about HTTP/1 (the consumers). Also thinking about streaming scenarios is interesting...

Anyways 😄

**Comment by stephentoub** [comment]
> I just think the people thinking about it are thinking about HTTP/1

Maybe some.  As you can see in the first paragraph of the cited https://github.com/stephentoub/corefx/commit/0e4d6401e76c79f8ef81b5064322b8de18a115a2, I'm not one of them: "We could add such APIs and make the 95% HTTP/1.1 case work fully synchronously without terribly impacting the code base.  But HTTP/2 is really hard."

**Comment by KrzysztofCwalina** [comment]
@ManickaP, I think it would be good to update the "motivation" section of this issue and add the main reason for this feature: there are lots of existing code bases that have very deep synchronous call stacks. Developers are simply not willing to rewrite these code bases to be asynchronous. If they need to call async only API in these synchronous methods, they use sync-over-async, which then in turn causes "soft" deadlock. We want to provide synchronous APIs for these developers because synchronous APIs, although inefficient, can can help in avoiding these soft deadlocks.

Of course, as you said usability is also a factor. We know that many developers prefer to prototype code using synchronous APIs. Synchronous APIs are better to prototyping. For example, they are easier to observe in the debugger.

**Comment by jkotas** [comment]
> If I create the potential for a thousand threads just in case

There would be at most one extra thread created for each synchronously waiting thread created by the application. So this should be reducing maximum number of threads by 2 at most. If the application creates 1000 threads, it has a bad performance problem in the first place and this won't fix it (maybe just move the needle a little bit).

> a thousand work items get queued that end up doing computationally and memory-intensive work

This is problem with async today as well. I do not think that the synchronous APIs like this should be advertised as a solution for the problem where CPU and memory-intensive work overwhelms the system and you I/O is not getting scheduled.

**Comment by stephentoub** [comment]
I think we're talking about different things.  My point was that in order to avoid deadlocking / incredibly long ramp up times any time there's a spike in traffic, app developers directly or indirectly employing sync-over-async calls need to set a very high min thread count for the thread pool, but such a min count isn't restricted to just work items that perform sync-over-async, but rather everything that's queued to the pool.

**Comment by benaadams** [comment]
> > They aren't sync

> How so?

Its unfortunate that we ended up with sync-vs-async as the terms. Synchronous is work done as same time; so the fast path of async, cpu bound tasks and other forms of IO that immediately return (data available).

.NET currently general has two forms of apis:

* "Blocking" where the entire thread is paused doing no useful work (its counter part being "non-blocking" which returns saying no data rather than blocking and can be use for polling; leading to high cpu use)

* "Async" which is the genius re-representation of callback/notification based code; to both allow it to be written as synchronous code and allow a fast path when it doesn't need to defer.

Both "Blocking" and "Async" are written as if they are synchronous code; whereas callback, non-blocking, push (Rx) need to be written differently, so are powerful representations because they look like everything else and don't unnecessarily consume cpu.

Calling "Async" "async" is fine as it describes the mechanism; however calling "Blocking" "synchronous" makes it sound more benign than it is; does not communicate the mechanism and more problematically suggests its something different than what it is.

So; to me, the phrases "sync-over-async" and "async-over-sync" while *I* know what the problems are, the problems are not immediately conveyed; and also raise the question is it problematic to do cpu-bound work (i.e. sync) in async method, which isn't the problem. 

If the phrase was "blocking-over-async" and "async-over-blocking"; it much more immediately clear what the problem is (the waiting mechanisms are in conflict). This is also coupled with the issue that the "blocking" apis seem simpler both in name and usage (historical PR is on their side).

It additionally becomes confusing when talking about async completing immediately; which is the "sync" fast-path (which is true); but has nothing to do with the "blocking" apis which are also referred to as "sync".

I don't necessarily think I'll gain traction in changing the terminology used; but it is my hill... 😄 

> > the first thing people will see is the "Send" method

> I don't think it is. Forget the alphabetical ordering, we're talking about 1 method on HttpClient among more than 30. 

I am happier with this restriction rather than providing "blocking" apis for everything

**Comment by Drawaes** [comment]
I can understand the "deep sync(avoiding the definition)" argument but then if you are moving a code base to .net core where these are missing maybe it's time to consider them as part of the move. Either way it's a valid case.

However the prototyping I am unsure of. If you make a console app with async Task as it's return type it's simply the await keyword, is there really a difference in that?

**Comment by stephentoub** [comment]
> Its unfortunate that we ended up with sync-vs-async as the terms.

FWIW, I don't think it's unfortunate, but reasonable people can differ.

My use of sync, async, and sync-over-async is entirely about when and how results are returned.  If you call a function and it returns the result to you, that's synchronous.  If you call a function and it returns to you without yet providing the result, that's asynchronous (whether you explicitly write a callback or you use a keyword to let the compiler generate the callback for you or you poll).  If you expose a sync method (it returns the answer synchronously), but you internally implement it by calling an asynchronous method and thus necessarily may block in order to wait for the answer to come, that's sync-over-async.

**Comment by darrelmiller** [comment]
It seems the default DelegatingHandler implementation of Send does a blocking call to SendAsync.  Doesn't that mean that any app developer that starts trying to call Send with an existing DelegatingHandler will end up doing Blocking over Async and not realize it?  Would it not be better to just throw a NotImplementedException if someone tries to call Send and the DelegatingHandler doesn't support Send.

**Comment by stephentoub** [comment]
> It seems the default DelegatingHandler implementation of Send does a blocking call to SendAsync.

That should not be the case. DelegatingHandler.Send should call InnerHandler.Send.

What is true is that the base HttpMessageHandler.Send will call SendAsync and block waiting for it.  All of the built-in HttpMessageHandler-derived types will override Send to do the "right thing", but other HttpMessageHandler-derived types will need to override Send, or else usage of HttpClient.Send will end up using their SendAsync implementations and blocking until they do.

**Comment by darrelmiller** [comment]
@stephentoub Yeah., my mistake, I was reading the wrong class.  It does call innerHandler.Send.  But if HttpMessageHandler then calls SendAsync and blocks, then the effect is the same.  This seems like there is a risk that devs are going to think they are getting pure Sync behaviour and that won't happen if unmodified classes that derive from DelegatingHandler are present in the pipeline.

**Comment by Drawaes** [comment]
That also is a very interesting point. We use delegating handlers (custom) for service discovery and for metrics collections. So even if you have a non sync over async method it will need to call the SendAsync on those won't it?

**Comment by stephentoub** [comment]
> This seems like there is a risk that devs are going to think they are getting pure Sync behaviour

Yup, it's one of the key concerns I previously called out.  But I think the alternative of having HttpMessageHandler.Send throw NotSupportedException is worse.

**Comment by itsmecurtis** [comment]
> > the first thing people will see is the "Send" method
> 
> I don't think it is. Forget the alphabetical ordering, we're talking about 1 method on HttpClient among more than 30. GetAync/PostAsync are much more likely to be the "first thing people will see".

Could the new sync `.Send` be an explicit implementation of a new interface (`IBlockingHttpClient`?) so that even among the more than 30 methods it will be clear that it supports a very narrow set of scenarios?

**Comment by KrzysztofCwalina** [comment]
> I can understand the "deep sync(avoiding the definition)" argument but then if you are moving a code base to .net core where these are missing maybe it's time to consider them as part of the move. Either way it's a valid case.

We don't want to add friction for people who want to move to .NET Core.

> However the prototyping I am unsure of. If you make a console app with async Task as it's return type it's simply the await keyword, is there really a difference in that?

The differences are small but non-trivial. We have quite consistent evidence from usability studies that the average .NET developer still struggles with async code. When code is already written and eveything works fine, async code looks simple and great. But for example, when people need to debug async code, the experience is not as great (and probably never will be), as I am sure you are aware. 

Also, there is a list of very small things that people run into when we watch them writing async code. Speaking about your console app example: when you create a console app in VS, the template is a synchronous main. Many people don't actually know how to convert it to async main. They add the async keyword and forget to change the return value to Task.

**Comment by jkotas** [comment]
> My point was that in order to avoid deadlocking / incredibly long ramp up times any time there's a spike in traffic, app developers directly or indirectly employing sync-over-async calls need to set a very high min thread count for the thread pool.

Yes, it is the case today. I belive that we can make it better by making the threadpool away of sync-over-async.

I am with @davidfowl that we need to see where we are going to end up eventually. It will take several years until this new synchronous API gets adoption. The cloud protocols are going to evolve in the meantime. For how long do we expect majority of the cloud / Azure to be on HTTP/1?

If the answer is that HTTP/1 started fading away or will start fading away soon, we need to start putting plan for HTTP/2 now to stay ahead of the curve. What would the plan for HTTP/2 or HTTP/3 be? Do the hard ugly rewrite? Or make threadpool aware of sync-over-async?

**Comment by Drawaes** [comment]
> Speaking about your console app example: when you create a console app in VS, the template is a synchronous main.

Time to change the template?

**Comment by stephentoub** [comment]
> I believe that we can make it better by making the threadpool away of sync-over-async.

I would like to understand the heuristics being proposed for that.  I agree it'd be excellent to improve things there, I've just yet to see a solution that addresses the issue while not having degenerate behavior in not-terribly-unreasonable situations.

> It will take several years until this new synchronous API gets adoption.

Azure SDK is planning to adopt it for when .NET 5 ships.  They already ship synchronous methods.

> What would the plan for HTTP/2 or HTTP/3 be? Do the hard ugly rewrite?

Currently blocking we do in the implementation could require up to N * 2 threads (for N requests).  With a rewrite / thoughtfulness I expect we could get it to N + 1 threads (with a hard/ugly rewrite).  I don't see any way to guarantee no more than N, in any implementation.

> For how long do we expect majority of the cloud / Azure to be on HTTP/1?

@KrzysztofCwalina? @Petermarcu?

**Comment by benaadams** [comment]
> > What would the plan for HTTP/2 or HTTP/3 be? Do the hard ugly rewrite?

> Currently blocking we do in the implementation could require up to N * 2 threads (for N requests). With a rewrite / thoughtfulness I expect we could get it to N + 1 threads (with a hard/ugly rewrite). I don't see any way to guarantee no more than N, in any implementation.

Aside https://github.com/stephentoub/corefx/commit/0e4d6401e76c79f8ef81b5064322b8de18a115a2 would only have an improvement on Windows for sync http/1.1 as on Linux sockets don't use sync IO (event for sync calls)

**Comment by stephentoub** [comment]
> as on Linux sockets don't use sync IO (event for sync calls)

This is not true.

**Comment by KrzysztofCwalina** [comment]
> For how long do we expect majority of the cloud / Azure to be on HTTP/1?

There are currently no non HTTP/1 services and none planned. Services that need persistent connections and duplex communication use AMQP

**Comment by benaadams** [comment]
Ah, my mistake, I thought it all went via epoll as it has handling for dealing with blocked events by signalling them rather than queuing to ThreadPool https://github.com/dotnet/runtime/blob/35a59c1d8debff1e62e7142b7bbaa22996ff689d/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L270-L278

**Comment by stephentoub** [comment]
Yeah, to do async we flip the socket to non-blocking, and there are potential race conditions that would result if we tried to flip it back, so if an async operation is performed, all subsequent sync operations are then emulated.

**Comment by davidfowl** [comment]
Appropriate label @scalablecory

**Comment by davidfowl** [comment]
Ah the other interesting scenario where this might  break down is blazor wasm running in the browser where there’s a single thread. We’ll need to decide what the behavior is going to be there (throw, hang, spin up a new thread?)

**Comment by stephentoub** [comment]
> Ah the other interesting scenario where this might break down is blazor wasm running in the browser where there’s a single thread.

How does this differ from HttpWebRequest.GetResponse? WebClient.DownloadString? Getting the response stream back from HttpClient and using its Read method?

**Comment by davidfowl** [comment]
![image](https://user-images.githubusercontent.com/95136/74590456-692d1280-4fc3-11ea-8e84-a7afe848a58a.png)

As of right now people aren't using those APIs in Blazor because it's new (sure some *old* libraries might be using them). We're adding a new way to hang the browser 😄.

Anyways, that's another environment where these sync methods aren't recommended. Something to call out as a concern.

**Comment by stephentoub** [comment]
Stream.Read is still very much alive. And works today with HttpClient.

> Anyways, that's another environment where these sync methods aren't recommended. 

Sure. The platform is chock-full of sync methods that perform IO. This is one more.

**Comment by davidfowl** [comment]
@stephentoub it is but not all APIs are equal (maybe worth making a list of APIs that do blocking IO). Every Blazor application that needs to talk to a server will use `HttpClient` to do so. 

Are we going to add blocking WebSocket APIs as well?

**Comment by stephentoub** [comment]
> Every Blazor application that needs to talk to a server will use HttpClient to do so.

And can happily use the plethora of async methods exposed.

> Are we going to add blocking WebSocket APIs as well?

There are no such plans. As you say, everyone needs to make HTTP requests. Few in comparison need to make websocket requests. And if you combine the everpresent need for HTTP with scenarios requiring sync, that drives a need for sync HTTP APIs that hasn't presented itself for websockets. On top of that, .NET has had sync HTTP APIs since day 1; it's never had sync websockets APIs. Demand for sync HTTP is simply much higher. We don't want to add these sync HTTP APIs for the heck of it; we feel compelled to based on customer demand.

**Comment by KrzysztofCwalina** [comment]
> Anyways, that's another environment where these sync methods aren't recommended.

As I already mentioned above, we need to have analyzers that will flag when a sync method is called in an async method's body. @davidfowl, don't you think it would pretty much eliminate the problem you brought up in practice?

**Comment by davidfowl** [comment]
@KrzysztofCwalina yes it should help mitigate some of the issues. But only if it’s on by default.

**Comment by KrzysztofCwalina** [comment]
> it should help mitigate some of the issues. But only if it’s on by default.

Yes, it should be on by default.

**Comment by davidfowl** [comment]
It might be good if that analyzer had some knowledge of frameworks as well so it could suggest the asynchronous alternative. If it only looked at the current method being an async method it would miss cases where you could define an MVC controller action has synchronous and then use the new HttpClient sync APIs safely and potentially cause thread pool starvation...

**Comment by terrajobst** [comment]
* Should we have an analyzer for flagging using sync methods if (1) the call site is in an async method and (2) there is an async version
* Can we promote async methods over sync methods in IntelliSense?

* `HttpClient`
    - The `Send` methods should mirror the structure of the `SendAsync` methods, as opposed to using defaulted parameters.
* There are no convenience methods
    - Such as `GetString`, but that's by-design because we want to discourage using these APIs for the broad set of consumer

```C#
namespace System.Net.Http
{
    // The main classes where we want the sync API.
    public partial class HttpMessageInvoker : IDisposable
    {
        // Existing async method
        public virtual Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);

        // Proposed new sync methods
        public virtual HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
    }
    public partial class HttpClient : HttpMessageInvoker
    {
        // Existing async methods
        public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request);
        public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption);
        public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken);
        public override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
        
        // Review suggestion: mirror async methods
        public HttpResponseMessage Send(HttpRequestMessage request);
        public HttpResponseMessage Send(HttpRequestMessage request, HttpCompletionOption completionOption);
        public HttpResponseMessage Send(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken);
        public override HttpResponseMessage SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
    }

    // The changes bellow enable proper implementation of the sync API.

    // HttpMessageHandler and derived classes
    public abstract partial class HttpMessageHandler : IDisposable
    {
        // Existing async method
        protected internal abstract Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken);
        
        // Proposed new sync method
        protected internal virtual HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
    }
    public abstract partial class DelegatingHandler : HttpMessageHandler
    {
        // Proposed new sync method
        protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
    }
    public abstract partial class HttpClientHandler : HttpMessageHandler
    {
        // Proposed new sync method
        protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
    }
    public abstract partial class SocketsHttpHandler : HttpMessageHandler
    {
        // Proposed new sync method
        protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
    }
    public abstract partial class MessageProcessingHandler : DelegatingHandler
    {
        // Proposed new sync method
        protected internal sealed override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken);
    }

    // HttpContent and derived classes
    public abstract partial class HttpContent : IDisposable
    {
        // Existing async methods
        public Task CopyToAsync(Stream stream);
        public Task CopyToAsync(Stream stream, TransportContext context);
        public Task CopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);
        public Task CopyToAsync(Stream stream, CancellationToken cancellationToken);

        protected abstract Task SerializeToStreamAsync(Stream stream, TransportContext context);
        protected virtual Task SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);

        // Proposed new sync methods
        public void CopyTo(Stream stream, TransportContext context, CancellationToken cancellationToken);

        protected virtual void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
    }
    public partial class ByteArrayContent : HttpContent
    {
        // Proposed new sync method
        protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
    }
    public partial class MultipartContent : HttpContent, IEnumerable<HttpContent>, IEnumerable
    {
        // Proposed new sync method
        protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
    }
    public sealed partial class ReadOnlyMemoryContent : HttpContent
    {
        // Proposed new sync method
        protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
    }
    public partial class StreamContent : HttpContent
    {
        // Proposed new sync method
        protected override void SerializeToStream(Stream stream, TransportContext context, CancellationToken cancellationToken);
    }
}
```

**Comment by jhudsoncedaron** [comment]
// Review suggestion: mirror async methods
        public HttpResponseMessage Send(HttpRequestMessage request);

Having actually gone the route of needing a sync HttpClient because deadlocks, we discovered this API is actually unweildy because it requires pregeneration of the entire request body. (The same would also be true of async code but you feel it a lot more with sync code.) We discovered that one of the these two API surfaces work better for requests with a body:

        public HttpResponseMessage Send(HttpRequestMessage request, Action<Stream> bodyGenerator);

or
        public HttpTransaction SendHeaders(HttpRequestMessage request);
        public class HttpTransaction /* nested class? */ : IDisposable() {
             Stream RequestBody { get; }
             HttpResponseMessage FinishRequest();
             void Dispose();
        }

In either case, the exposed stream is a direct wrapper around the network stream and actually uploads the request as you write to it. This requires HTTP/1.1 and chunked transfer encoding so the body doesn't have to have a size. Note that any request verb other than HEAD may have a body.

**Comment by KrzysztofCwalina** [comment]
@jhudsoncedaron, can your problem be solved with a custom or stream based HttpContent, i.e. the generation of the body happens in HttpContent.CopyTo implementation?

**Comment by jhudsoncedaron** [comment]
@KrzysztofCwalina : There's no way to say "Don't call `CreateContentReadStream`".

**Comment by davidsh** [comment]
>@KrzysztofCwalina : There's no way to say "Don't call CreateContentReadStream".

WCF has lots of use cases where they create custom HttpContent object's that are stream based are similar to your scenarios.

**Comment by stephentoub** [comment]
@jhudsoncedaron, I don't understand the problem.  You just need a custom HttpContent, e.g. something along the lines of:
```C#
internal class ActionHttpContent
{
    private readonly Action<Stream> _action;
    public ActionHttpContent(Action<Stream> action) => _action = action;
    protected override void SerializeToStream(Stream stream) => _action(stream);
    ...
}
...
var message = new HttpRequestMessage(...) { Content =  new ActionHttpContent(s =>
{
    ... // write the body to s here
}) };
```

**Comment by normj** [comment]
What is the trigger for adding this now? From AWS we have been telling users since .NET Core 1.0 days we don't support sync operations in our SDK for the .NET Standard version unlike the .NET Framework version because the HttpClient didn't support sync and best practices as published by Microsoft was to use async. With this being added that messaging is muddled and there is going to be pressure put on libraries authors to support a bad pattern. Most likely any intellisense or analyzers created to encourage use of async pattern won't automatically apply for libraries that are built around HttpClient. It just seems like a surprising move now after all of the work over the last couple years to make async all over the place including async mains I would have understand doing it when .NET Standard 2.0 was being created with a goal to increase compatibility.

**Comment by jhudsoncedaron** [comment]
@stephentoub : Perhaps it does, but I wasn't able to determine it does from the documentation alone.

**Comment by karelz** [comment]
@normj see answers above that sum it up:
- https://github.com/dotnet/runtime/issues/32125#issuecomment-585348061 (usage in sync components/SDKs)
- https://github.com/dotnet/runtime/issues/32125#issuecomment-585491076 (usability, prototyping)

> Why now?

Because now we have the feedback. Loud and clear. It's been coming for quite some time and it reached intensity we cannot simply ignore anymore.

> Most likely any intellisense or analyzers created to encourage use of async pattern won't automatically apply for libraries that are built around HttpClient.

Libraries can ship their own analyzers if they choose so, or plug in into existing ones.
Libraries have to opt-in into synchronous APIs and have to make choice to expose them as well. If they do (for the reasons above), they should also warn users about compromise on perf in high-perf scenarios in such cases (e.g. in docs).

We do not want to encourage mass usage of sync APIs and that's why we do not expose all convenience methods from `HttpClient`.

**Comment by epignosisx** [comment]
Considering .NET Framework only had `WebClient` and `WebRequest` to make synchronous HTTP calls. Could we keep these proposed `HttpClient` sync APIs internal and use them under the hood for `WebClient` and `WebRequest`?

This should unblock customers using those APIs since they can continue to use them and do not have the risk of deadlocks.

**Comment by karelz** [comment]
`WebRequest` and `WebClient` are legacy APIs which are not even fully implemented (on purpose) - key trouble with them is that they use process-wide configuration model, which makes components in applications impact each other more than they should. Overall they caused too much trouble over time.
We do not plan to improve them, unless there is strong customer signal they block migration to .NET Core. They are even more suboptimal on perf than sync APIs (until recently they created connection per request and in some scenarios they still do).

**Comment by normj** [comment]
@karelz Sorry I'm trying to avoid a debate over Azure vs AWS SDK but this enhancement just has a bad smell to it. 

Like you said the feedback had been coming for a while but I believe the platform took the correct choice and tried to raise the quality of code being produced for .NET and pushed its developers to write better higher performant code. Then it looks like Azure comes in and says this is too hard let’s go back on the principals we stood for since the beginning of .NET Core. Which that puts other library authors in the awkward position do they support sync and build analyzers to discourage their use or not support them and have their libraries deemed not as usable as the other libraries that choose the bad design. 

I’m not naïve about the business driver but that does not paint a great picture of .NET being its own independent product driven by the community if core features just get changed because another Microsoft team wants it.

**Comment by KrzysztofCwalina** [comment]
It's not that "another Microsoft team wants it". The dynamics we are seeing is the following: there are orders of magnitude more existing applications than new application being written. So far, these old "legacy" applications were mainly left in maintenance mode and so we could have only modern great APIs. But the Cloud is maturing an enterprises suddenly want to port more and more of these old apps to the cloud. When they port these apps to Azure, let's say port a piece of code that used to write to a file (using sync APIs), and now needs to write to Azure Storage, they notice that the storage APIs don't have synchronous APIs. This adds friction to the porting effort, and in majority of cases forces people to use sync-over-async, which then in turn cause soft deadlocks. First class synchronous APIs would remove the friction and let people avoid the soft deadlock in most cases. 

And so, I would not characterize what's happening as one Microsoft team forcing the platform to do the wrong thing, but rather, the platform adopting to what customers want, i.e. they want to port their existing apps to environments where where networking is more common, without the friction associated with rewriting a synchronous app to an asynchronous app.

Lastly, I think I would be more careful with pushing for these synchronous APIs, if the rest of the APIs involved in network communication were sync only, i.e. if the world was indeed pure async, I would not want to destroy the purity. But the purity some people allude to above does not actually exist. People today call async Send, block for the result, and then use synchronous read APIs on the response stream. Usually there are more calls to the synchronous stream APIs than the initial Send, and so stopping people from calling the sync Send does not buy us anything. It just adds friction (dissatisfaction).

**Comment by karelz** [comment]
Another angle comes from user studies -- which demonstrate that most developers are confused by async APIs, don't know how to use them and rather give up :(. It is sad eye opener, and a thing to consider.
Most other platforms do have synchronous APIs, which are great for prototyping, experimenting and even ok for majority of production environments (except the top high-performance environments). It is competitive disadvantage for .NET to not have HTTP sync APIs.

All these things together is what changed our view point. I am personally not super happy about it, but if it is what our customers need, we should meet them where they are.

**Comment by darrelmiller** [comment]
Not giving a sync option removes the need for developers to make the decision on whether they should call Send or SendAsync.  We have been telling developers for years that  they should use async calls for things that wait for IO . Now we are saying sometimes they shouldn't.  Our answer to when they shouldn't use async seems to be "when it is too hard" or for a specific performance scenario where sync ends up being better, for reasons that went over my head.

Providing a choice is just an opportunity for more developers to make bad choices.  It is curious that usability studies are showing that most developers are confused by async APIs.  Aren't all JavaScript HTTP calls async?  Is this a problem in that ecosystem too?

It also seems like if we are trying to make this easier for developers who can't get async to work, then having just the single Send method and not all the other overloads is going to be extremely frustrating.
Imagine explaining that you can do  

    var result = await httpClient.GetStringAsync("https://bing.com");

but if you can't get async to work for you, then it's ok, you can just do

    var response =  httpClient.Send(new HttpRequestMessage() { RequestUrl = new Uri("https://bing.com")};
    var result = response.Content.ReadAsString();
 
This seems to make the "for prototyping"  use case a bit weak.

**Comment by benaadams** [comment]
> Most other platforms do have synchronous APIs, which are great for prototyping, experimenting and even ok for majority of production environments (except the top high-performance environments). It is competitive disadvantage for .NET to not have HTTP sync APIs.

Node and JavaScript thrive without any sync apis and have wholeheartily embraced `async` keyword (and Rx); and people are very happy to prototype with them.

Is there anything that can be learnt from how `async` is more approachable and understandable by JS/node devs and bring that to .NET? It must be more than there just being no sync apis?

**Comment by karelz** [comment]
> @darrelmiller It also seems like if we are trying to make this easier for developers who can't get async to work

Primarily we are enabling higher-level libraries to deliver easier developer experiences (sync) without forcing them to use sync-over-async.
We do not optimize for direct `HttpClient` usage. We should also document the `Send` API drawbacks: no support for HTTP/2+, potential problem with custom Handlers, etc.

**Comment by karelz** [comment]
@benaadams we are aware of the JavaScript pure async nature. If you have insights what we could learn and how to bring it to .NET, we are all ears ...

**Comment by darrelmiller** [comment]
> We should also document the Send API drawbacks: no support for HTTP/2+
That's all you needed to say.  :-) Now it will be easy for me to explain to people why they should only use the async methods

**Comment by scalablecory** [comment]
> Not giving a sync option removes the need for developers to make the decision on whether they should call Send or SendAsync. We have been telling developers for years that they should use async calls for things that wait for IO . Now we are saying sometimes they shouldn't. Our answer to when they shouldn't use async seems to be "when it is too hard" or for a specific performance scenario where sync ends up being better, for reasons that went over my head.

This concern was brought up during API review. See @terrajobst's [API review comment](https://github.com/dotnet/runtime/issues/32125#issuecomment-591016000) for the video. One possibility we discussed was to make Intellisense understand async, and depending on the user's method being sync or async, to show the sync or async versions of the method first. We understand it is not ideal, but it is also not a new problem: most of the other async APIs .NET have this same issue.

> That's all you needed to say. :-) Now it will be easy for me to explain to people why they should only use the async methods

The addition of this API does not change recommendations of when async vs sync is appropriate.

**Comment by jkotas** [comment]
> Send API drawbacks: no support for HTTP/2+

The API is going work for HTTP/2+ using sync-over-async fallback. Is that correct? I would not describe it as no support for HTTP/2+. HTTP/2+ is going to be supported with asterisks.

**Comment by stephentoub** [comment]
> The API is going work for HTTP/2+ using sync-over-async fallback. Is that correct?

Correct. We already do that when synchronously reading from the response stream, and this will extend it to cover the request issuing as well.

**Comment by jhudsoncedaron** [comment]
> sync-over-async fallback

I'm having trouble following this conversation. Assuming this means calling Task.Result or Task.GetAwaiter().GetResult() from inside the sync method (which is what the term meant elsewhere); that was one of the drivers that lead for me to abandon HttpClient. (The deadlock wasn't a soft deadlock; we had to bounce the server manually a few times.)

**Comment by stephentoub** [comment]
> I'm having trouble following this conversation. Assuming this means calling Task.Result or Task.GetAwaiter().GetResult()

- When you use the async methods on HttpClient, all I/O is performed asynchronously. The implementation never blocks synchronously on a Task waiting for it to complete.
- Same goes for the async methods on the returned response Stream.
- Stream also has synchronous methods. If you use those on the response stream, with HTTP/1.x we will end up calling through to the synchronous methods on Socket. With HTTP/2, we don't have that ability, and may end up doing the equivalent of task.Wait() while waiting for the shared connection loop to provide response data.
- With this proposal, none of that changes. In addition to the synchronous methods on the response Stream, there are then also synchronous methods to initiate the request. And the same conditions apply as with the response stream... for HTTP/1.x, all synchronous operations end up in synchronous calls to the sync methods on the underlying Socket, and for HTTP/2, we may end up in the equivalent of task.Wait.

**Comment by davidfowl** [comment]
> Stream also has synchronous methods. If you use those on the response stream, with HTTP/1.x we will end up calling through to the synchronous methods on Socket. With HTTP/2, we don't have that ability, and may end up doing the equivalent of task.Wait() while waiting for the shared connection loop to provide response data.

Aren't responses buffered by default?

**Comment by tmds** [comment]
>    That approach has a long history of AVs. Have those been dealt with?
Does it? I hope so!

https://github.com/dotnet/corefx/pull/38804 deals with them. The change is not in any released version of .NET Core.

These are the known limitations:

https://github.com/dotnet/runtime/blob/40e44e7ed07071bd2606b51cf0ac7071a23e3075/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs#L201-L204

**Comment by stephentoub** [comment]
> Aren't responses buffered by default?
At the handler level, no. Any buffering of the response Stream that's layered on by HttpClient is subject to the exact same conditions I described for the response Stream. HttpClient.GetStreamAsync just returns that same Stream, no buffering. GetString/ByteArrayAsync buffer out of necessity. Get/SendAsync buffer to support the HttpCompletionOption.ResponseContentRead option, which is the default for those methods when it's not specified.

**Comment by davidfowl** [comment]
I meant at the HttpClient level, I also thought GetStreamAsync was buffered unless `HttpCompletionOption.ResponseHeadersRead` was used.

For HTTP/2, maybe it makes sense to degrade the sync API to make parallel requests to the same connection spin up a new connection since multiplexing over a singe connection is broken anyways?

**Comment by stephentoub** [comment]
> I also thought GetStreamAsync was buffered unless HttpCompletionOption.ResponseHeadersRead was used.

GetStreamAsync always specifies ResponseHeadersRead; it doesn't provide an alternative. :)

> For HTTP/2, maybe it makes sense to degrade the sync API to make parallel requests to the same connection spin up a new connection since multiplexing over a singe connection is broken anyways?

Interesting idea. My hope/expectation is that https://github.com/dotnet/runtime/pull/32917 will be sufficient / strike the right balance.

**Comment by davidfowl** [comment]
> GetStreamAsync always specifies ResponseHeadersRead; it doesn't provide an alternative. :)

Right.

> Interesting idea. My hope/expectation is that #32917 will be sufficient / strike the right balance.

It's definitely better than spawing another thread to complete the blocked one. 

What about async requests queued behind synchronous requests? In particular, what happens if the request that establishes the connection is synchronous and then another async requests is made after it? I guess I could read the code but I figure it's worth bringing up these weird interactions now...

**Comment by stephentoub** [comment]
> what happens if the request that establishes the connection is synchronous and then another async requests is made after it?

Same thing as is they were both async: the latter will asynchronously wait for the connection to be established. For the first request, we could choose to make the connection established synchronously or have it be sync-over-async, but either way the subsequent async request will asynchrously wait for the shared connection to be established.

**Comment by GeraudFabien** [comment]
Maybe there could be an external library with the extention methods to add sync to httpclient. I wouldn't put them in .Net for many reason:
-  If you put in .net for every method async you will have to do an sync method. 
   - When i was a beginner it was the fact that some sync method exist that make me lost a lot of time. The fact they exist made me believe that i must'n use async method on a sync environment (that why this thread impact me). If the common case exist in sync but not a peculiar case you will thinks that it's a complex case to handle or a bad practice and will be lost
   - You will have a lot of issue open to ask you for
- If you encounter a bug and has to do a lot of work to handle a very uncommon case. If it's not a .net feature you may not handle it or just raise exception. If it's the framework you will have to fix it

There is also the others solutions : 
- To document httpclient pages to show how to make the call syncronously. 
- Add to intellisence the cleverness that if a method is sync it propose a code snippet 
- Add in the documentation of async method a link to a way to do it syncronously

Even after reading the thread i still thinks this is for :
- legacy app that will not pass to net framework because they have no budget sync httpclient or not. I highly doubt that sync call will be even discuss since it's only maintenance
- People that cannot use async lib. I never accounter one. Even the people that never touch C# find really quickly how to use it. (They start by just calling .Result most of the times). Like benaadams stat e JS on the other hand as just async. I add to that no way to easly found how to await synchronously (We never found it so i don't even know if it's possible).
- Framework that don't agree with a best practice that exist for 8 years !!!
- Small app for me is not even a good point since it's not more complex to do async await on small app. I agree debug on async await can be harder be it really happend on complex app. When you're app do one or two request in a foreach and you have one layer of async it's unlikely to be hard to debug

**Comment by jhudsoncedaron** [comment]
@GeraudFabien : Sometimes you cannot use an async call because you're implementing an interface that isn't async. Note that `.Result` and `.GetAwaiter().GetResult()` can deadlock if there's an async method above you on the stack. This isn't about normal complexity problems anymore. This problem hasn't gone away because sometimes you really can't use async because the context doesn't allow it. It is impossible to build an async application that calls a sync library that turns around and calls an async library, and this particular API draws a lot of sync libraries trying to use it.

**Comment by jhudsoncedaron** [comment]
> Add in the documentation of async method a link to a way to do it syncronously

I've followed three different recommendations on this, they were all wrong. The real answer is it's impossible to run an async method synchronously.

**Comment by valeriob** [comment]
I'll add my point of view to the motivation.
There are many AspNet MVC 5 applications that where born before async was a thing and are actively developed today (think tens of mloc), we can bring them to aspnet core, and find out that the same code suffer from thread starvation.

**Comment by GeraudFabien** [comment]
@jhudsoncedaron My main point was there are other solution thant implemented it in the runtime. Documentation can point to other implementation that cover the need and are syncronous. I didn't use it for years but https://github.com/restsharp/RestSharp/wiki/Getting-Started isn't async (and seems to be the only real sync client library). There is also https://github.com/jeffijoe/httpclientgoodies.net https://flurl.dev/ that are async but allow to read simply the value.

@valeriob yes. My point was did your client give you money to migrate you're application. Wether it was yes or no in both case it's not just because of async in Httpclient that you make you're choice. But on the ROI for your projet and the global risk taken. In a nutshell can you promise me that if we add for sync in the native HttpClient will you migrate *all* you're legacy to dotnet core. 

Sync api are just a very very small problem to the migration. The fear of breaking change will block most of the application to migrate until they have a really good reason  to. And when they will have a reason to they may consider thinks like nuget change, project.json, tooling upgrade, learning curve, ... Before even thinking of the lack of sync api on httpclient.

**Comment by davidfowl** [comment]
> There are many AspNet MVC 5 applications that where born before async was a thing and are actively developed today (think tens of mloc), we can bring them to aspnet core, and find out that the same code suffers of thread starvation.

The thread pool will still starve, because you're still blocking a thread pool thread. It'll just happen slightly slower than if you didn't have the sync APIs.

**Comment by valeriob** [comment]
> The thread pool will still starve, because you're still blocking a thread pool thread. It'll just happen slightly slower than if you didn't have the sync APIs.

Since sync over async will consume 2 threads instead of one, that's the core of the problem, am i right ? I'm not sure how it will translate in the performance characteristics of the application.

**Comment by valeriob** [comment]
> @valeriob yes. My point was did your client give you money to migrate you're application. Wether it was yes or no in both case it's not just because of async in Httpclient that you make you're choice. But on the ROI for your projet and the global risk taken. In a nutshell can you promise me that if we add for sync in the native HttpClient will you migrate _all_ you're legacy to dotnet core.

You are absolutly right @GeraudFabien, i've migrated many small applications already, and a big one. It all depends on how heavy is the usage of http apis. If your database has an http interface, you will starve pretty quickly when switching to aspnet core, if you use EntityFramework for example, it still have sync api, so the application will behave the same and no change needed.

Anyway my point is, if you do have a sync codebase with heavy usage of httpclient, the client today has to do sync over async, so you need to raise the threadpool limits pretty quickly (we had to do it as soon as we went from MVC to aspnetcore) to survive.
So now there are 2 options going forward : 
1) Migrate the codebase to async (pretty expensive)
2) If http client sync api lands, and library autors adopt it you gain instant benefits

For example, this is the code that a library has to implement to allow sync apis on top of httpclient async ones, but maybe it's just too late, the cat is out of the bag :D 
https://github.com/ravendb/ravendb/blob/9b5313689310f4938c0c2f9a4385e9fad2708cfe/src/Raven.Client/Util/AsyncHelpers.cs

It would be interesting to hear the opinion of libraries autors that are close to customers, for example is it just me suffering from having a sync solution using databases over http ? 😄

**Comment by KrzysztofCwalina** [comment]
> The thread pool will still starve, because you're still blocking a thread pool thread. It'll just happen slightly slower than if you didn't have the sync APIs.

I thought starvation is different and better than the "soft deadlock" we are talking about. A sync app, at some point, will run out of thread-pool threads and so work will be advancing slowly, but it will be advancing. In sync-over-async case, there are fairness issues between the i/o completion threads and threads picking up new work, and so the app might not be able to make progress. At least this is what I understood after talking to @stephentoub.

**Comment by stephentoub** [comment]
> I thought starvation is different and better than the "soft deadlock" we are talking about here.

Yes.  Here's a (maybe-overly-simplified-but-gets-the-point-across) example.  Comment out just one of Sync/SyncOverAsync/Async and run it:
```C#
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

class Program
{
    static void Main()
    {
        var sw = Stopwatch.StartNew();
        Sync();
        //SyncOverAsync();
        //Async();
        Console.WriteLine(sw.Elapsed);
    }

    static void Sync() => Task.WaitAll(Enumerable.Range(0, Environment.ProcessorCount * 10).Select(_ => Task.Run(() => Thread.Sleep(100))).ToArray());
    static void SyncOverAsync() => Task.WaitAll(Enumerable.Range(0, Environment.ProcessorCount * 10).Select(_ => Task.Run(() => Task.Delay(100).Wait())).ToArray());
    static void Async() => Task.WaitAll(Enumerable.Range(0, Environment.ProcessorCount * 10).Select(_ => Task.Run(async () => await Task.Delay(100))).ToArray());
}
```
I get results like this:
```
Sync:          00:00:01.1129462
SyncOverAsync: 00:00:08.1362978
Async:         00:00:00.1313958
```

**Comment by jhudsoncedaron** [comment]
@KrzysztofCwalina The following is actually a hard deadlock. I've hit it.

    async ValueTask<string> Method() {
        await Task.Delay(1);
        using (var h = new HttpClient())
            return h.DownloadStringAsync("some url");
    }

Everybody's saying it's a soft deadlock. But sometimes the thread pool can't make new threads. Everything grinds to a halt.

(Nobody is disputing the correct fix is to use `return await` here. Try it with an intervening interface like IReadOnlyDictionary<> and you'll find out you can't.)

**Comment by stephentoub** [comment]
> Everybody's saying it's a soft deadlock. But sometimes the thread pool can't make new threads.

On .NET Core?  Please share a repro.

**Comment by jhudsoncedaron** [comment]
@stephentoub : We ran out of nonpaged memory due to overload. It took awhile to get any diagnostics because Remote Desktop was dead meat and nothing was being logged.

**Comment by stephentoub** [comment]
> We ran out of nonpaged memory due to overload.

Sure, if you run out of memory, that's an entirely different ball of wax.

**Comment by casperOne** [comment]
While I dislike the idea of introducing a sync mechanism, I understand the need to address real pain points that people are expressing.

However, everything I've seen in this thread suggests adding methods to `HttpClient`.  Has it been considered to introduce a _completely new_ class which exposes the sync methods?

I like this over an interface on `HttpClient` or adding the sync methods to `HttpClient`:

- It seems that everyone is in agreement that sync methods should only be used in limited situations (prototyping or high performance scenarios); encapsulation in another class would be a great warning sign saying "beware all ye who enter here"
- Analyzers would have an easier time determining if sync code is used in async code; when putting it all on `HttpClient` we'll certainly need an attribute to point to other methods for each async/sync method pair.  If it's tucked away in another class, it's much neater (and an attribute would be able to target the sync/async version very accurately because we can use `typeof` in the attribute)
- This allows library authors much more freedom (and doesn't impose restrictions on sync/async support being on the same class) in determining _how_ they want to support async _and_ sync (although I suspect many will not support it at all).

**Comment by jhudsoncedaron** [comment]
Considering on analysis that we shouldn't use the same instance of HttpClient for both sync and async calls (need to disable HTTP/2 for sync calls to really be sync), casperOne is most likely correct. Too bad this idea appeared so late.

**Comment by kekyo** [comment]
Exactly agreed too, casperOne. I feel people will confuse and can't understand how to implement better asynchronous and scalability for safer. (In my experience, they can't understand what is topic for 'logical Task' at now and the interface will mislead)

**Comment by valeriob** [comment]
I've been looking at the httpclient sync api, i'm surprised about the implementation, it's just a GetAwaiter().GetResult() behind a method ? how does this help anybody suffering from thread pool starvation ?

https://github.com/ManickaP/runtime/blob/7844fbf39812bd591dc10dc0fdfbd5fc1137d507/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L464

**Comment by stephentoub** [comment]
> I've been looking at the httpclient sync api, i'm surprised about the implementation, it's just a GetAwaiter().GetResult() behind a method ? how does this help anybody suffering from thread pool starvation ?

See the Debug.Assert on the previous line.  That GetAwaiter().GetResult() isn't blocking because the task will have completed synchronously.  The GetResult() is just there to propagate any exceptions.

**Comment by valeriob** [comment]
Thanks @stephentoub

**Comment by rellis-of-rhindleton** [comment]
I know this is closed and I don't want to beat a dead horse... but for the benefit of anybody still wondering why we would want a sync Send method, here's one reason.

In ASP.NET Core (or any app that uses IHostBuilder I assume) you can use `ConfigureAppConfiguration` to add your own custom configuration providers. Unfortunately this arrangement does not support asynchronous operations. So if you write your own custom provider that, for example, queries a config server over HTTP, that provider has to run synchronously.

A better solution to that particular problem would be to support asynchronous configuration providers. But I can't find that support today.
